### DISCUSSION TITLE
Support Ethereum-style JSON-RPC interface

### DISCUSSION THREAD

**Discussion by ychiaoli18** [filecoin-project/FIPs#422] [Enhancements - Technical]
## Motivations

FVM’s M2.1 aims to support the EVM and the existing EVM ecosystem out of the box. Currently, the EVM’s existing developer tools and decentralized apps utilize Ethereum-style JSON-RPC APIs to interact with blockchains. Filecoin’s implementations, such as Lotus or Forest, do not speak those APIs now. For the existing EVM-based tools to support Filecoin with minimal change, Filecoin’s implementations have to implement the Ethereum-style JSON-RPC APIs.

This discussion aims to provide a prioritized list of the JSON-RPC methods that Filecoin should implement. We researched how Hardhat, Remix, and Alchemy use each method to understand their importance. The methods are ranked based on the following criteria:

- Nodes are not expected to hold keys, so the methods requiring the nodes to have private key information are excluded, such as eth_accounts, eth_signTransaction, eth_sign, etc.
- Nodes are not expected to provide any mining capabilities, so the methods related to mining are excluded, such as eth_hashrate, eth_getWork, eth_submitWork, etc.
- Nodes are not expected to provide information about uncle blocks, so methods like eth_getUncleByBlockHashAndIndex are excluded.
- Methods used by Hardhat and Remix are a high priority.
- Methods used by logs and events are a medium priority. Part of the reason is that the implementation detail of logs and events on FVM is currently under discussion.
- The rest of the methods are a low priority.
- Complexity is provided based on the FVM team’s experience on Lotus.

### Included Methods

The methods are sorted according to the order of implementation priority. 

| Method                                  | Importance | Complexity | Description                                                                                                                                        |
| --------------------------------------- | ---------- | ---------- | -------------------------------------------------------------------------------------------------------------------------------------------------- |
| eth_blockNumber                         | High       | Low        | Returns the number of most recent block.                                                                                                           |
| eth_getBlockByHash                      | High       | Low        | Returns information about a block by hash.                                                                                                         |
| eth_getBlockByNumber                    | High       | Low        | Returns information about a block by number.                                                                                                       |
| eth_getTransactionByHash                | High       | Low        | Returns the information about a transaction requested by transaction hash.                                                                         |
| eth_getTransactionCount                 | High       | Low        | Returns the number of transactions sent from an address.                                                                                           |
| eth_getTransactionReceipt               | High       | Low        | Returns the receipt of a transaction by transaction hash.                                                                                          |
| eth_getTransactionByBlockHashAndIndex   | High       | Low        | Returns information about a transaction by block hash and transaction index position.                                                              |
| eth_getTransactionByBlockNumberAndIndex | High       | Low        | Returns information about a transaction by block number and transaction index position.                                                            |
| eth_sendRawTransaction                  | High       | Low        | Submits a raw transaction.                                                                                                                         |
| eth_getBalance                          | High       | Low        | Returns the balance of the account of given address.                                                                                               |
| eth_gasPrice                            | High       | Low        | Returns the current price per gas in wei.                                                                                                          |
| eth_chainId                             | High       | Low        | Returns the chain ID of the current network.                                                                                                       |
| net_version                             | High       | Low        | Returns the current network id.                                                                                                                    |
| net_listening                           | High       | Low        | Returns true if client is actively listening for network connections.                                                                              |
| eth_getCode                             | High       | Medium     | Returns code at a given address.                                                                                                                   |
| eth_call                                | High       | High       | Executes a new message call immediately without creating a transaction on the block chain.                                                         |
| eth_subscribe                           | High       | High       | Ethereum API - Subscribe to different event types like newHeads, logs, and pendingTransactions using websockets.                                   |
| eth_unsubscribe                         | High       | High       | Subscriptions are cancelled with a regular RPC call with eth_unsubscribe as method and the subscription id as first parameter.                     |
| eth_estimateGas                         | High       | Not sure   | Generates and returns an estimate of how much gas is necessary to allow the transaction to complete.                                               |
| eth_getStorageAt                        | Medium     | Medium     | Returns the value from a storage position at a given address.                                                                                      |
| eth_getLogs                             | Medium     | High       | Returns an array of all logs matching filter with given id.                                                                                        |
| eth_getFilterChanges                    | Medium     | High       | Polling method for a filter, which returns an array of logs which occurred since last poll.                                                        |
| eth_getFilterLogs                       | Medium     | High       | Returns an array of all logs matching filter with given id.                                                                                        |
| eth_newBlockFilter                      | Medium     | High       | Creates a filter in the node, to notify when a new block arrives.                                                                                  |
| eth_newFilter                           | Medium     | High       | Creates a filter object, based on filter options, to notify when the state changes (logs).                                                         |
| eth_newPendingTransactionFilter         | Medium     | High       | Creates a filter in the node, to notify when new pending transactions arrive.                                                                      |
| eth_uninstallFilter                     | Medium     | High       | Uninstalls a filter with given id.                                                                                                                 |
| eth_feeHistory                          | Medium     | High       | Transaction fee history                                                                                                                            |
| eth_getBlockTransactionCountByHash      | Low        | Low        | Returns the number of transactions in a block from a block matching the given block hash.                                                          |
| eth_getBlockTransactionCountByNumber    | Low        | Low        | Returns the number of transactions in a block matching the given block number.                                                                     |
| eth_protocolVersion                     | Low        | Low        | Returns the current Ethereum protocol version.                                                                                                     |
| eth_getBlockReceipts                    | Low        | Medium     | Get all transaction receipts for a given block on Ethereum.                                                                                        |
| eth_getProof                            | Low        | Not sure   | Returns the merkle proof for a given account and optionally some storage keys.                                                                     |
| eth_maxPriorityFeePerGas                | Low        | Not sure   | Returns a fee per gas that is an estimate of how much you can pay as a priority fee, or "tip", to get a transaction included in the current block. |

_Note: description field is adapted from [Ethereum's API Spec](https://github.com/ethereum/execution-apis) and [Alchemy's API Doc](https://docs.alchemy.com/alchemy/apis/ethereum)_

### Excluded Methods

| Method                            | Description                                                                                           |
| --------------------------------- | ----------------------------------------------------------------------------------------------------- |
| eth_accounts                      | Returns a list of addresses owned by client.                                                          |
| eth_sign                          | Returns an EIP-191 signature over the provided data.                                                  |
| eth_signTransaction               | Returns an RLP encoded transaction signed by the specified account.                                   |
| eth_sendTransaction               | Signs and submits a transaction.                                                                      |
| eth_getUncleByBlockHashAndIndex   | Returns information about an uncle of a block by hash and uncle index position.                       |
| eth_getUncleByBlockNumberAndIndex | Returns information about an uncle of a block by number and uncle index position.                     |
| eth_getUncleCountByBlockHash      | Returns the number of uncles in a block from a block matching the given block hash.                   |
| eth_getUncleCountByBlockNumber    | Returns the number of transactions in a block matching the given block number.                        |
| eth_syncing                       | Returns an object with data about the sync status or false.                                           |
| eth_coinbase                      | Returns the client coinbase address.                                                                  |
| eth_mining                        | Returns whether the client is actively mining new blocks.                                             |
| eth_hashrate                      | Returns the number of hashes per second that the node is mining with.                                 |
| eth_getWork                       | Returns the hash of the current block, the seedHash, and the boundary condition to be met (“target”). |
| eth_submitWork                    | Used for submitting a proof-of-work solution.                                                         |
| eth_submitHashrate                | Used for submitting mining hashrate.                                                                  |

**Comment by mriise** [comment]
Events are still in spec land: https://github.com/filecoin-project/fvm-specs/issues/98

**Comment by Stebalien** [comment]
> eth_sendRawTransaction

This has High complexity as we'll need to re-wrap the transaction as a Filecoin message (for vaildiation via an abstract ETH account: https://github.com/filecoin-project/FIPs/discussions/388).
